[search ui] Add asset filter fields to search index

clairelin135 opened the pull request:
This PR enables adds the following asset filters to the search index results:
- asset owner
- compute kind
- code location
- asset group

This involves:
1. Querying for these fields per-asset on `SECONDARY_SEARCH_QUERY`
2. Grouping by field to determine the # of assets per filter
3. Adding each filter to the list of possible search results

Open questions:
- Perf impact of fetching these additional fields for each asset in graphQL?

clairelin135 commented:
* **#20373** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20373?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#20372** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20372?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @clairelin135 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

clairelin135 commented on js_modules/dagster-ui/packages/ui-core/src/search/types.ts:
```diff
@@ -14,12 +14,33 @@ export enum SearchResultType {
+  AssetGroup = 'AssetFilterSearchResultType.AssetGroup',
+}
+
+export function isAssetFilterSearchResultType(
```
Not sure how to check if a value exists in a given enum... I was seeing the following always return true:
```
x: SearchResultType | AssetFilterSearchResultType = ...
x in AssetFilterSearchResultType // Returns true??
```
For now, just tested for strict equivalence to make this logic work

clairelin135 replied:
Got it, thanks.

I think we still need this logic within the asset search component to render asset results and filter results a little differently.

clairelin135 replied:
I moved this function to the other PR where it's actually used

hellendag reviewed (dismissed):
Some thoughts inline about separating the search components/hooks, related to comments on https://github.com/dagster-io/dagster/pull/20373.

clairelin135 commented on js_modules/dagster-ui/packages/ui-core/src/search/SearchDialog.tsx:
```diff
@@ -118,17 +124,25 @@ export const SearchDialog = ({searchPlaceholder}: {searchPlaceholder: string}) =
+      }
     },
     [searchSecondary],
   );
```
Thanks for the explanation here.

I played around with pulling this callback out into a hook, though I realized after your other comment about avoiding building unwanted results that we could configure whether asset filters are returned at the `useGlobalSearch` hook level.

After that change, this filtering logic is no longer needed.

clairelin135 commented on js_modules/dagster-ui/packages/ui-core/src/search/SearchDialog.tsx:
```diff
@@ -118,17 +124,25 @@ export const SearchDialog = ({searchPlaceholder}: {searchPlaceholder: string}) =
+        dispatch({
+          type: 'complete-secondary',
+          queryString: queryStringForResults,
+          results: results.filter((result) => result.item.type === SearchResultType.Asset), // Only return asset results
```
Good call, I implemented that within `useGlobalSearch`. A flag now is provided to `useGlobalSearch` to indicate whether asset filter results should be returned or not.

clairelin135 commented:
@hellendag appreciate the explanations here.

I addressed the feedback and now `SearchDialog` doesn't need to know about asset filtering behavior, I think it's ready for another look!

salazarm commented on js_modules/dagster-ui/packages/ui-core/src/assets/AssetsOverview.tsx:
```diff
@@ -49,7 +48,21 @@ type CountPerCodeLocation = {
 };
 
-function buildAssetCountBySection(assets: AssetTableFragment[]): AssetCountsResult {
+type AssetDefinitionMetadata = {
+  definition: {
+    owners: Array<
+      {__typename: 'UserAssetOwner'; email: string} | {__typename: 'TeamAssetOwner'; team: string}
+    >;
+    computeKind: string | null;
+    groupName: string | null;
+    repository: {
+      name: string;
+      location: {name: string};
+    };
+  } | null;
+};
+
```
I would go with something like this in order to keep a single source of truth for the base types.

```
type AssetDefinitionMetadata = {
  definition: Pick<AssetNode['definition'],  'owners' | ' computeKind' | 'groupName' | 'repository'>;
```
or if you don't like picking keys like that you could do 
```
type AssetDefinitionMetadata = {
  definition: {
    owners: AssetNode['definition']['owners']
    ...
  },
```

clairelin135 replied:
good call, I added this

salazarm commented:
One down side to this approach is that since there can be 2 instances of `useGlobalSearch` and they don't share cache with each other which means they will separately query the same data (via apollo so they should hit the shared cache if done one after the other, though it's possible both are open together in which case the cache isn't available so absent any apollo client query deduping it would be a duplicate request. afaik apollo doesnt dedupe).

salazarm commented:
I guess for now it's probably fine and we can optimize that later if it becomes a problem.

clairelin135 commented on js_modules/dagster-ui/packages/ui-core/src/search/useGlobalSearch.tsx:
```diff
@@ -174,6 +259,11 @@ type IndexBuffer = {
+// This is the version of the secondary query, used as part of the cache key.
+// When the data in the cache must be invalidated, this version should be bumped to prevent
+// fetching stale data.
+export const SEARCH_SECONDARY_DATA_VERSION = 'v1;';
```
We recently added caching to store the latest asset query result to load the search UI when the query hasn't completed. This PR adds additional fields to that query, but this causes an issue on the first load as the logic assumes those fields exist but they don't on the cached query data. This causes the asset search UI to be unloadable.

This PR adds a version to the key to ensure that we won't fetch stale data.

salazarm commented on js_modules/dagster-ui/packages/ui-core/src/search/useGlobalSearch.tsx:
```diff
@@ -207,7 +297,7 @@ export const useGlobalSearch = () => {
   } = useIndexedDBCachedQuery<SearchSecondaryQuery, SearchSecondaryQueryVariables>({
     query: SEARCH_SECONDARY_QUERY,
-    key: 'SearchSecondary',
+    key: `SearchSecondary:${SEARCH_SECONDARY_DATA_VERSION}`,
```
This works but it leaves the old cache in place so it never gets cleaned up. I think instead `useIndexedDBCachedQuery` should take a third parameter `version` that can be used to tell `useIndexedDBCacheQuery` to ignore caches on a different version.

clairelin135 replied:
Makes sense. This PR is updated to now also store the data version in the cache.

I added a data version for both the primary and secondary queries since we should always be providing a version

salazarm reviewed (approved):
Sweet
